File: C/lt_01_estseq/lt01_estseq_30.c
#include <stdio.h>

int makeDias(int mes) {
  switch (mes) {
  case 1:
    mes = 31;
    break;
  case 2:
    mes = 59;
    break;
  case 3:
    mes = 90;
    break;
  case 4:
    mes = 120;
    break;
  case 5:
    mes = 151;
    break;
  case 6:
    mes = 181;
    break;
  case 7:
    mes = 212;
    break;
  case 8:
    mes = 243;
    break;
  case 9:
    mes = 273;
    break;
  case 10:
    mes = 304;
    break;
  case 11:
    mes = 334;
    break;
  case 12:
    mes = 365;
    break;
  default:
    mes = 0;
    break;
  }
  return mes;
}

int makeMes(int dia) {
  int mes;

  if (dia >= 365) {
    printf("Erro Mais ou 365 Dias enviados para makeMes");
  } else if (dia >= 334) {
    mes = 11;
  } else if (dia >= 304) {
    mes = 10;
  } else if (dia >= 273) {
    mes = 9;
  } else if (dia >= 243) {
    mes = 8;
  } else if (dia >= 212) {
    mes = 7;
  } else if (dia >= 181) {
    mes = 6;
  } else if (dia >= 151) {
    mes = 5;
  } else if (dia >= 120) {
    mes = 4;
  } else if (dia >= 90) {
    mes = 3;
  } else if (dia >= 59) {
    mes = 2;
  } else if (dia >= 31) {
    mes = 1;
  } else {
    mes = 0;
  }
  return mes;
}

int main() {
  int dia, mes, ano, nascimento, atual, biss;

  do {
    printf("Insira a data de Nascimento em DD:MM:YY : ");
    scanf("%d:%d:%d", &dia, &mes, &ano);

    nascimento = ano * 365;
    nascimento += dia;
    nascimento += makeDias(mes);
    biss = ((ano / 4) - (ano / 100) + (ano / 400));
    nascimento += biss;
  } while (nascimento <= 0);

  do {
    printf("Insira a data Atua em DD:MM:YYYY : ");
    scanf("%d:%d:%d", &dia, &mes, &ano);

    atual = ano * 365;
    atual += dia;
    atual += makeDias(mes);
    atual += ((ano / 4) - (ano / 100) + (ano / 400));
  } while ((nascimento > atual) || (atual <= 0));

  biss = ((ano / 4) - (ano / 100) + (ano / 400)) - biss;
  atual = atual - biss - nascimento;

  ano = atual / 365;
  atual = atual % 365;
  mes = makeMes(atual);
  atual -= makeDias(mes);
  dia = atual;

  printf("Tem %.2d Anos %.2d Messes e %.2d Dias de Idade.\n", ano, mes, dia);
  return 0;
}

File: rust/lt01-estseq/lt01-estseq-30.rs
use std::io;

fn main() {
    // input
    let mut dn = input_i32("Insira a dia de nascimento: ");
    let mn = input_i32("Insira o mês de nascimento: ");
    let an = input_i32("Insira o ano de nascimento: ");
    let mut da = input_i32("Insira a dia de atual: ");
    let mut ma = input_i32("Insira o mês de atual: ");
    let mut aa = input_i32("Insira o ano de atual: ");
    //Calc atual
    da += aa * 365;
    da += (aa / 4) - (aa / 100) + (aa / 400);
    da += mes_dia(ma);
    //Calc nascimento
    dn += an * 365;
    dn += (an / 4) - (an / 100) + (an / 400);
    dn += mes_dia(mn);
    // Calc resulstado
    da -= dn;
    aa = da / 365;
    da %= 365;
    ma = dia_mes(da);
    da -= mes_dia(ma);
    // print
    println!(
        "A Idade da pessoa é {} anos {} messes e {} dias de idade.",
        aa, ma, da
    );
}

fn mes_dia(mes: i32) -> i32 {
    match mes {
        1 => return 0,
        2 => return 31,
        3 => return 90,
        4 => return 120,
        5 => return 151,
        6 => return 181,
        7 => return 212,
        8 => return 243,
        9 => return 273,
        10 => return 304,
        11 => return 334,
        12 => return 365,
        _ => return 0,
    }
}

fn dia_mes(mut dia: i32) -> i32 {
    let mut i = 0;
    loop {
        if i < 7 {
            if i == 1 && dia >= 28 {
                i += 1;
                dia -= 28;
            } else if i % 2 == 0 && dia >= 31 {
                i += 1;
                dia -= 31
            } else if i % 2 == 1 && dia >= 30 {
                i += 1;
                dia -= 30;
            } else {
                return i;
            }
        } else {
            if i % 2 == 0 && dia >= 30 {
                i += 1;
                dia -= 30
            } else if i % 2 == 1 && dia >= 31 {
                i += 1;
                dia -= 31;
            } else {
                return i;
            }
        }
    }
}

fn input_i32(texto: &str) -> i32 {
    let int: i32;
    loop {
        println!("{}", texto);
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Error read_line");
        int = match input.trim().parse() {
            Ok(n) => {
                if n >= 0 {
                    n
                } else {
                    println!("Valor inválido!");
                    continue;
                }
            }
            Err(_) => {
                println!("Valor inválido!");
                continue;
            }
        };
        return int;
    }
}
